app/javascript/packs/followButtons.js

```diff
@@ -0,0 +1,346 @@
+/* global showLoginModal userData */
+
+/**
+ * Helper function to check if a user is currently following a tag
+ *
+ * @param {Object} user The current user
+ * @param {string} tagId The ID of the tag
+ *
+ * @returns {boolean}
+ */
+const doesUserFollowTag = (user, tagId) =>
+  !!JSON.parse(user.followed_tags).find((tag) => tag.id === tagId);
```

Comment:
Refactor (blocking): Run `JSON.parse` once, most likely as a variable that `doesUserFolloTag` will call find on because Otherwise it will be parsed `n` times in the loop lower down in the code.

```javascript
  nonUserFollowButtons.forEach((button) => {
    const { info, fetched } = button.dataset;
    const buttonInfo = JSON.parse(info);
    if (
      buttonInfo.className === 'Tag' &&
      user &&
      button.dataset.buttonInitialized !== 'true'
    ) {
      const initialButtonFollowState = doesUserFollowTag(user, buttonInfo.id)
        ? 'true'
        : 'false';
      updateInitialButtonUI(initialButtonFollowState, button);
      button.dataset.buttonInitialized = true;
    } else if (fetched !== 'fetched') {
      fetchFollowButtonStatus(button, buttonInfo);
    }
  });
```

And actually, this is a double loop, so O(n^2), because we have `nonUserFollowButtons.forEach` and then in each loop of that we're calling `Array.prototype.find`. If the initial `JSON.parse` could create a `Set` instead, we would have faster lookup. Something like this maybe:

```javascript
const followedTagIds = new Set(JSON.parse(user.followed_tags).map(({ tagId }) => tagId))
```

And then we probably don't even need this function anymore. We can do the following:

```javascript
  nonUserFollowButtons.forEach((button) => {
    const { info, fetched } = button.dataset;
    const buttonInfo = JSON.parse(info);
    if (
      buttonInfo.className === 'Tag' &&
      user &&
      button.dataset.buttonInitialized !== 'true'
    ) {
      const initialButtonFollowState = followedTagIds.has(buttonInfo.id)
        ? 'true'
        : 'false';
      updateInitialButtonUI(initialButtonFollowState, button);
      button.dataset.buttonInitialized = true;
    } else if (fetched !== 'fetched') {
      fetchFollowButtonStatus(button, buttonInfo);
    }
  });
```